app/src/main/java/com/willowtree/vocable/room/VocableDatabaseMigrations.kt

```diff
@@ -53,4 +54,59 @@ object VocableDatabaseMigrations {
             database.execSQL("ALTER TABLE Phrase_New RENAME TO Phrase")
         }
     }
+
+    val MIGRATION_3_4: Migration = object : Migration(3, 4) {
+        // Moving to new JSON schema
+        override fun migrate(database: SupportSQLiteDatabase) {
+            // Create new Category, Phrase, & Cross Ref tables
+            database.execSQL("CREATE TABLE Category_New (category_id TEXT NOT NULL, creation_date INTEGER NOT NULL, is_user_generated INTEGER NOT NULL, resource_id INTEGER, localized_name TEXT, hidden INTEGER NOT NULL, sort_order INTEGER NOT NULL, PRIMARY KEY(category_id))")
+            database.execSQL("CREATE TABLE Phrase_New (phrase_id TEXT NOT NULL, creation_date INTEGER NOT NULL, is_user_generated INTEGER NOT NULL, last_spoken_date INTEGER NOT NULL, resource_id INTEGER, localized_utterance TEXT, sort_order INTEGER NOT NULL, PRIMARY KEY(phrase_id))")
+            database.execSQL("CREATE TABLE CategoryPhraseCrossRef_New (category_id TEXT NOT NULL, phrase_id TEXT NOT NULL, PRIMARY KEY(category_id, phrase_id))")
+
+            // Get My Sayings
+            val categoryId = PresetCategories.USER_FAVORITES.id
+            val crossRefCursor =
+                database.query("SELECT phrase_id FROM CategoryPhraseCrossRef WHERE category_id = '$categoryId'")
+            val myLocalizedSayings = LinkedHashSet<String>()
+            val phraseIds =  mutableListOf<String>()
+            while (crossRefCursor.moveToNext()) {
+                val phraseId = crossRefCursor.getString(crossRefCursor.getColumnIndex("phrase_id"))
+                phraseIds.add(phraseId)
+            }
+            crossRefCursor.close()
+
+            phraseIds.forEach {
+                val phraseCursor = database.query("SELECT localized_utterance FROM Phrase WHERE phrase_id = '$it'")
+                while (phraseCursor.moveToNext()) {
+                    val saying = phraseCursor.getString(phraseCursor.getColumnIndex("localized_utterance"))
+                    myLocalizedSayings.add(saying)
+                }
+                phraseCursor.close()
+            }
+
+            // If we didn't pick up any sayings from scheme 3, check if there were some in
+            // scheme 2
+            val nonLocalizedSayings = VocableSharedPreferences().getMySayings()
```

Comment:
Do we need this check here? I think that if the user is being migrated from 2 to 4, Room will run 2 to 3 then 3 to 4

Reply:
I don't think the actual presets repo initialization will happen before 3-4 runs so we would only wind up with the old scheme in mySayings